# **Pandas**

# **What is Pandas ?**
 ***The pandas is library in python .which is used for data loading and data manipulating and data analysis.***

- There are two types of data structure mainly supported by pandas**
 
 **1.Series**
      - 1-D array(list,tuple,set,dictionary).
 
 **2.DataFrame**
      - 2-D table like excel work sheet.

# How to install pandas ?
- use pip install pandas as pd

**import library**

In [2]:
import pandas as pd   # pandas use for data loading and data manipulating.

**Load the Dataset.**

In [4]:
df=pd.read_csv("C:\\Users\\Abc\\Downloads\\pandas_practice1_dataset.csv") # to load the data set


**1.head()** : ***To get first 5 rows in dataset.***

In [6]:
df.head() # to display the first 5 rows.

Product_ID Product_Name     Brand     Category    Price  Stock_Units  \
0        1001       Router  Logitech  Accessories   574.15        188.0   
1        1002       Laptop        HP  Accessories  1644.89        203.0   
2        1003        Mouse  Logitech  Peripherals   642.71        325.0   
3        1004     Keyboard   Samsung  Peripherals   888.96        439.0   
4        1005     Keyboard     Canon  Electronics  1500.41         51.0   

  Country_Of_Origin Launch_Date  
0           Germany  2020-06-22  
1                UK  2023-10-15  
2             India  2019-02-22  
3               USA  2020-12-22  
4             Japan  2019-11-27

**2.tail.()** : ***To get last 5 rows in dataset.***

In [8]:
df.tail() # to display the  last 5 rows.

Product_ID Product_Name   Brand     Category    Price  Stock_Units  \
510        1433      Monitor      HP  Peripherals   318.13         48.0   
511        1069      Printer  Lenovo  Peripherals   823.61        234.0   
512        1072   Smartphone    Asus  Electronics  1854.61        363.0   
513        1351     Keyboard   Apple  Accessories   801.33        340.0   
514        1123   Headphones  Lenovo  Accessories  1162.56        283.0   

    Country_Of_Origin Launch_Date  
510             India  2018-12-19  
511             China  2023-04-10  
512             India  2023-05-06  
513             India  2021-09-28  
514             Japan  2023-06-26

**3.describe(include='all')** : ***To summarize the data both numerical and texual column.***

In [10]:
df.describe(include='all') # to describe the data both categorical and numerical

Product_ID Product_Name  Brand     Category        Price  \
count    515.000000          515    495          515   495.000000   
unique          NaN           10      8            3          NaN   
top             NaN       Laptop  Apple  Peripherals          NaN   
freq            NaN           63     67          182          NaN   
mean    1249.976699          NaN    NaN          NaN  1027.038485   
std      144.491687          NaN    NaN          NaN   574.943081   
min     1001.000000          NaN    NaN          NaN    52.680000   
25%     1123.500000          NaN    NaN          NaN   526.785000   
50%     1251.000000          NaN    NaN          NaN  1026.990000   
75%     1375.500000          NaN    NaN          NaN  1563.940000   
max     1500.000000          NaN    NaN          NaN  1994.630000   

        Stock_Units Country_Of_Origin Launch_Date  
count    494.000000               515         515  
unique          NaN                 6         440  
top             NaN             India  2023-05-06  
freq            NaN                94           4  
mean     256.018219               NaN         NaN  
std      144.328485               NaN         NaN  
min        0.000000               NaN         NaN  
25%      128.750000               NaN         NaN  
50%      258.000000               NaN         NaN  
75%      379.750000               NaN         NaN  
max      497.000000               NaN         NaN

**Product_Name**
- count : The total count of product name is 515  this show there is no missing value in this column.
- unique : In product Name 10 unique values are present.
- top : The Laptop is most repeated  name  in the column.
- frequency : The laptop is 63 times are repeated.

**Brand**
- count : The total count of Brand is 495  this show there  is some missing values in this column.
- unique : In Brand 8 unique values are present.
- top : The Apple is most repeated  name  in the column.
- frequency : The Apple brand laptop is 67 times are repeated.

**Category**
- count : The total count of Brand is 515  this show there is no missing values in this column.
- unique : In category 3 unique values are present.
- top : The Peripherals	 is most repeated  name  in the column.
- frequency : The Peripherals is 182 times are repeated.

**Country_Of_Origin**
- count : The total count of is 515  this show there is no missing values in this column.
- unique : In category 6 unique values are present.
- top : The India is most repeated  name  in the column.
- frequency : The Peripherals is 94 times are repeated.


**4.describe() : Statistically summarize the data frame (i.e mean,max,min,count,standard deviation,top,unique).**

In [13]:
df.describe() # to summary of numerical column

Product_ID        Price  Stock_Units
count   515.000000   495.000000   494.000000
mean   1249.976699  1027.038485   256.018219
std     144.491687   574.943081   144.328485
min    1001.000000    52.680000     0.000000
25%    1123.500000   526.785000   128.750000
50%    1251.000000  1026.990000   258.000000
75%    1375.500000  1563.940000   379.750000
max    1500.000000  1994.630000   497.000000

**Price**
- The  price column has 495 non-null entries i.e some price values are missing. 
- mean : The average (mean) price of product is 1027.04 i.e  average ,product are priced around 1027 .
- std : The standard deviation is 547.94  is less than mean i.e the data  values are closer to mean, or around the mean.
- min : The smallest value is 52.
- max : The largest value is 1994.
- 25 % (Q1):  The 25 % value is 526  that show the values below the 25 %  of the data fall.
- 50 % (median): The middle value of this price column is 1026.99.
- 75 % (Q3) : 75 % of the product  are price at or below 1563.94   with only the high 25 % price above this.

**Stock unit**
- count : The stock unit column has 494 non null entries,i.e some stock values are missing.
- mean : The  average is 256 that show the on average ,product have about 256 unit available.
- std : The standard deviation is 144 which is less than mean i.e the data  values are closer to mean, or around the mean.
- min : The smallest  stock is 0.
- 25 % (Q1) : The  value is 128  that show the values below the 25 %  of the data fall.
- 50 % (median): The middle value of stock unit column is 258.
- 75 % (Q3) : 75 % of the product  are price at or below 497  with only the high 25 % price above this.

**5.shape** : ***To give total number of rows and columns.***

In [16]:
df.shape # shows the number of rows and column

(515, 8)

- In our dataset there is 515 rows and 8 columns are present.

**6.info()**: ***summarize the dataframe (i.e shows the column_name,non_null and datatypes).***

In [19]:
df.info() # shows the summary of dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 515 entries, 0 to 514
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Product_ID         515 non-null    int64  
 1   Product_Name       515 non-null    object 
 2   Brand              495 non-null    object 
 3   Category           515 non-null    object 
 4   Price              495 non-null    float64
 5   Stock_Units        494 non-null    float64
 6   Country_Of_Origin  515 non-null    object 
 7   Launch_Date        515 non-null    object 
dtypes: float64(2), int64(1), object(5)
memory usage: 32.3+ KB


- In above product_ID  Data type is Integer. 
- Product_Name,Brand ,Category,Country_of_origin,Launch_Date have object  datatype.
- In Brand,Price,stock unit  some missing values are present.
- price and stock unit have float data type.

**7.nunique()**: ***It return the total number of unique values in column.***

In [22]:
df.nunique() # that shows the unique values in data

Product_ID           500
Product_Name          10
Brand                  8
Category               3
Price                480
Stock_Units          311
Country_Of_Origin      6
Launch_Date          440
dtype: int64

**In productID 500,In Product name 10 ,Brand 8, Category 3,price 480,stock units 311 ,country of origin 6 , launch date 440 unique values are available.**

**8.isnull().sum()** :***It check the  missing values and count them.***

In [25]:
df.isnull().sum() # isnull() to check for missing values and sum() to get the total count of missing values  

Product_ID            0
Product_Name          0
Brand                20
Category              0
Price                20
Stock_Units          21
Country_Of_Origin     0
Launch_Date           0
dtype: int64

- In productID ,Product name,category,country of origin and launch date there is no missing values are present.
- In Brand ,price, stock units have some missing values.

In [48]:
df['Company']=df['Company'].fillna(df['Company'].mode()[0]) # to fill missing value in brand column
df['Price']=df['Price'].fillna(df['Price'].mean()) # fill missing value in price column
df['Stock_Units']=df['Stock_Units'].fillna(0) # fill missing value in stock unit column

In [50]:
df.isnull().sum()

Product_ID           0
Product_Name         0
Company              0
Category             0
Price                0
Stock_Units          0
Country_Of_Origin    0
Launch_Date          0
dtype: int64

In [30]:
df.nunique()

Product_ID           500
Product_Name          10
Brand                  8
Category               3
Price                480
Stock_Units          311
Country_Of_Origin      6
Launch_Date          440
dtype: int64

In [32]:
df['Brand'].nunique() # to get unique value in specific column

8

- There are 8 unique values in Brand column.

**9.duplicated()** : ***To check duplicate value in column.***

In [34]:
df.duplicated().sum() # shows the duplicate in dataset .sum()  total count of duplicates

15

- There are 15 duplicates values in our dataset.

**10.rename()**: ***To change the column name.***

In [36]:
df.rename(columns={'Brand':'Company'},inplace=True) # rename/change the column name

**11.columns** : ***To return the column names.***

In [38]:
df.columns # to get column names

Index(['Product_ID', 'Product_Name', 'Company', 'Category', 'Price',
       'Stock_Units', 'Country_Of_Origin', 'Launch_Date'],
      dtype='object')

In [40]:
df['Company'].unique() # shows the unique values in Company column

array(['Logitech', 'HP', 'Samsung', 'Canon', 'Apple', 'Asus', 'Lenovo',
       'Dell', nan], dtype=object)

**12.value_counts()**: ***To count the unique value in each column.***

In [42]:
df['Company'].value_counts() # shows the value count in each category

Company
Apple       67
Lenovo      66
Samsung     64
Dell        63
Asus        62
HP          58
Canon       58
Logitech    57
Name: count, dtype: int64

- There are 67 product from Apple.
- 66 entries from Lenovo.
- 64 product from samsung.
- also 63 value from Dell,62 product from Asus,58 product from HP,58 product from Canon and 57 product values from Logitech company . that show unique values in each company.

**13.groupby()** : ***To create a group. or Aggregate data by category.***

In [44]:
df.groupby("Company")["Price"].mean() # to create group

Company
Apple       1085.020615
Asus         983.353729
Canon       1016.422807
Dell        1111.049833
HP          1156.624727
Lenovo       955.801148
Logitech     994.607593
Samsung      961.732813
Name: Price, dtype: float64

**14.sort_values()** : ***Sorting the rows in ascending or descending oreder.***

In [46]:
df.sort_values(by="Price",ascending=False) # sort the price by ascending order

Product_ID Product_Name   Company     Category    Price  Stock_Units  \
44         1045       Laptop     Apple  Electronics  1994.63        323.0   
504        1122       Webcam        HP  Accessories  1987.56        110.0   
121        1122       Webcam        HP  Accessories  1987.56        110.0   
482        1483        Mouse     Canon  Electronics  1983.72        447.0   
184        1185       Webcam     Canon  Electronics  1979.61        386.0   
..          ...          ...       ...          ...      ...          ...   
372        1373     Keyboard  Logitech  Peripherals      NaN        260.0   
440        1441      Monitor      Asus  Accessories      NaN        181.0   
447        1448   Smartphone      Dell  Electronics      NaN        243.0   
470        1471     Keyboard    Lenovo  Accessories      NaN        208.0   
486        1487   Headphones      Asus  Accessories      NaN         56.0   

    Country_Of_Origin Launch_Date  
44            Germany  2023-12-30  
504           Germany  2020-09-13  
121           Germany  2020-09-13  
482               USA  2020-08-19  
184               USA  2022-04-06  
..                ...         ...  
372                UK  2023-08-13  
440                UK  2019-10-13  
447             Japan  2019-09-10  
470               USA  2022-07-29  
486             Japan  2022-09-13  

[515 rows x 8 columns]

- In above column the highest price is 1994.63.

**15.drop()** : ***To remove row or column.***

In [55]:
df.drop("Product_ID",axis=1,inplace=True) # drop() is used to remove the column

In [57]:
df.drop(0,axis=0,inplace=True)  # dropping rows using index position,axis=0 for rows

**16.index**: ***Return the row index.***

In [59]:
df.index

RangeIndex(start=1, stop=515, step=1)

In [71]:
df[df['Price'] > 1500] # filtring the data where price > 1500

Product_Name Company     Category    Price  Stock_Units Country_Of_Origin  \
1         Laptop      HP  Accessories  1644.89        203.0                UK   
4       Keyboard   Canon  Electronics  1500.41         51.0             Japan   
5         Router   Apple  Electronics  1575.61        256.0             India   
8     Smartphone  Lenovo  Electronics  1548.60        238.0             China   
10        Laptop      HP  Peripherals  1663.45        267.0                UK   
..           ...     ...          ...      ...          ...               ...   
504       Webcam      HP  Accessories  1987.56        110.0           Germany   
506       Webcam    Asus  Peripherals  1733.11        113.0               USA   
507       Webcam    Dell  Electronics  1655.76        106.0                UK   
508        Mouse    Dell  Peripherals  1824.90        108.0           Germany   
512   Smartphone    Asus  Electronics  1854.61        363.0             India   

    Launch_Date  
1    2023-10-15  
4    2019-11-27  
5    2021-06-13  
8    2020-04-07  
10   2022-03-10  
..          ...  
504  2020-09-13  
506  2018-05-08  
507  2023-04-11  
508  2020-05-20  
512  2023-05-06  

[146 rows x 7 columns]

- This shows the all price which is greater than 1500.

In [76]:
df[df['Company'] =='HP']

Product_Name Company     Category        Price  Stock_Units  \
1         Laptop      HP  Accessories  1644.890000        203.0   
10        Laptop      HP  Peripherals  1663.450000        267.0   
21    Headphones      HP  Electronics  1589.320000        432.0   
24    Smartphone      HP  Electronics  1490.420000        353.0   
28      Keyboard      HP  Electronics  1520.430000        439.0   
46        Router      HP  Accessories   182.270000        124.0   
67      Keyboard      HP  Accessories   230.420000        289.0   
87       Monitor      HP  Accessories  1027.038485        321.0   
89       Monitor      HP  Accessories   412.160000         80.0   
96    Smartphone      HP  Peripherals  1496.580000        412.0   
97    Smartphone      HP  Peripherals   768.430000         58.0   
102       Laptop      HP  Accessories  1653.200000        297.0   
103       Router      HP  Electronics  1677.620000         11.0   
105   Smartphone      HP  Peripherals  1402.560000         54.0   
116       Laptop      HP  Accessories  1320.400000         48.0   
121       Webcam      HP  Accessories  1987.560000        110.0   
137      Monitor      HP  Peripherals   592.230000        131.0   
151   Smartphone      HP  Accessories  1373.700000         78.0   
154       Webcam      HP  Electronics  1659.980000          0.0   
172     Keyboard      HP  Peripherals  1050.270000        360.0   
174       Router      HP  Accessories  1828.380000         24.0   
178       Webcam      HP  Accessories   251.120000        365.0   
180      Printer      HP  Electronics   634.880000         64.0   
199       Router      HP  Accessories   198.800000        212.0   
219     Keyboard      HP  Electronics   103.530000        341.0   
226       Laptop      HP  Electronics  1525.420000        359.0   
228   Headphones      HP  Peripherals   941.290000         47.0   
247       Tablet      HP  Peripherals  1913.800000        128.0   
249   Headphones      HP  Peripherals  1581.380000        485.0   
260        Mouse      HP  Accessories   226.710000        496.0   
264      Printer      HP  Accessories   272.110000         33.0   
272   Headphones      HP  Accessories  1880.320000          3.0   
274       Tablet      HP  Electronics  1759.900000        115.0   
285       Webcam      HP  Electronics  1027.038485        170.0   
288      Printer      HP  Peripherals   904.720000        435.0   
298        Mouse      HP  Peripherals  1948.060000        441.0   
299        Mouse      HP  Peripherals  1225.540000        105.0   
300   Smartphone      HP  Peripherals  1928.660000        281.0   
304      Printer      HP  Accessories  1484.430000        352.0   
306      Printer      HP  Accessories  1373.570000        412.0   
329       Laptop      HP  Peripherals   664.420000        290.0   
346       Webcam      HP  Accessories  1027.038485         28.0   
352       Laptop      HP  Accessories   130.120000        256.0   
355       Webcam      HP  Peripherals  1082.250000         49.0   
364   Smartphone      HP  Peripherals   431.530000        146.0   
383        Mouse      HP  Electronics   799.760000        313.0   
385     Keyboard      HP  Accessories  1265.570000        194.0   
389       Tablet      HP  Peripherals   178.300000        207.0   
417        Mouse      HP  Accessories  1849.900000          0.0   
431      Printer      HP  Electronics  1597.680000        248.0   
432      Monitor      HP  Peripherals   318.130000         48.0   
442       Tablet      HP  Electronics  1039.190000          0.0   
446   Headphones      HP  Accessories  1626.790000        216.0   
456        Mouse      HP  Peripherals  1690.250000        331.0   
458       Router      HP  Accessories  1208.010000        349.0   
490       Tablet      HP  Peripherals  1728.260000        384.0   
504       Webcam      HP  Accessories  1987.560000        110.0   
510      Monitor      HP  Peripherals   318.130000         48.0   

    Country_Of_Origin Launch_Date  
1             

- This show all record of HP company.

In [79]:
df[df['Price'] > 1500].shape[0]# filtering the column base on conditon and shap[0]  that count the each product

146

- In this data there are 146 count.

**17.loc[]** : ***Access by label.***

In [82]:
df.loc[3] # loc is use for label based indexing.returns rows with the index label 3

Product_Name            Keyboard
Company                  Samsung
Category             Peripherals
Price                     888.96
Stock_Units                439.0
Country_Of_Origin            USA
Launch_Date           2020-12-22
Name: 3, dtype: object

In [85]:
df.loc[:,["Product_Name","Price"]] # this is shows only Selected column  values

Product_Name    Price
1         Laptop  1644.89
2          Mouse   642.71
3       Keyboard   888.96
4       Keyboard  1500.41
5         Router  1575.61
..           ...      ...
510      Monitor   318.13
511      Printer   823.61
512   Smartphone  1854.61
513     Keyboard   801.33
514   Headphones  1162.56

[514 rows x 2 columns]

**18.iloc[]** : ***Access by position.***

In [87]:
df.iloc[2] # returns the row for the specified index position . 

Product_Name            Keyboard
Company                  Samsung
Category             Peripherals
Price                     888.96
Stock_Units                439.0
Country_Of_Origin            USA
Launch_Date           2020-12-22
Name: 3, dtype: object

In [90]:
df.iloc[0:10] # accessing the records for all the  columns with the index 0-10

Product_Name   Company     Category    Price  Stock_Units  \
1        Laptop        HP  Accessories  1644.89        203.0   
2         Mouse  Logitech  Peripherals   642.71        325.0   
3      Keyboard   Samsung  Peripherals   888.96        439.0   
4      Keyboard     Canon  Electronics  1500.41         51.0   
5        Router     Apple  Electronics  1575.61        256.0   
6        Tablet      Asus  Peripherals   168.36        304.0   
7        Laptop     Apple  Peripherals   897.22        493.0   
8    Smartphone    Lenovo  Electronics  1548.60        238.0   
9    Headphones   Samsung  Peripherals   135.93         95.0   
10       Laptop        HP  Peripherals  1663.45        267.0   

   Country_Of_Origin Launch_Date  
1                 UK  2023-10-15  
2              India  2019-02-22  
3                USA  2020-12-22  
4              Japan  2019-11-27  
5              India  2021-06-13  
6                USA  2021-11-13  
7                USA  2023-04-26  
8              China  2020-04-07  
9              India  2022-06-18  
10                UK  2022-03-10

In [94]:
df['Additional_price']=df['Price']+ 5
df['Additional_price']    # to add in 5 each price and create new column

1      1649.89
2       647.71
3       893.96
4      1505.41
5      1580.61
        ...   
510     323.13
511     828.61
512    1859.61
513     806.33
514    1167.56
Name: Additional_price, Length: 514, dtype: float64

- This add 5 Rs in in each price.and create a new column.

In [97]:
df.sort_values(by="Additional_price",ascending=False)

Product_Name Company     Category    Price  Stock_Units Country_Of_Origin  \
44        Laptop   Apple  Electronics  1994.63        323.0           Germany   
504       Webcam      HP  Accessories  1987.56        110.0           Germany   
121       Webcam      HP  Accessories  1987.56        110.0           Germany   
482        Mouse   Canon  Electronics  1983.72        447.0               USA   
184       Webcam   Canon  Electronics  1979.61        386.0               USA   
..           ...     ...          ...      ...          ...               ...   
39    Headphones  Lenovo  Peripherals    63.92        263.0             China   
382      Monitor    Asus  Peripherals    61.24        327.0             China   
398       Laptop   Apple  Peripherals    60.53        231.0             India   
255       Webcam    Asus  Accessories    57.36        330.0                UK   
375      Monitor    Dell  Electronics    52.68        437.0             Japan   

    Launch_Date  Additional_price  
44   2023-12-30           1999.63  
504  2020-09-13           1992.56  
121  2020-09-13           1992.56  
482  2020-08-19           1988.72  
184  2022-04-06           1984.61  
..          ...               ...  
39   2020-08-19             68.92  
382  2021-05-26             66.24  
398  2019-07-07             65.53  
255  2021-07-17             62.36  
375  2019-12-30             57.68  

[514 rows x 8 columns]

**19.rank()** : ***Rank value in a column.***

In [99]:
df['Additional_price'].rank(ascending=False) # rank to change position and return value

1      100.0
2      361.0
3      295.0
4      146.0
5      122.0
       ...  
510    438.5
511    309.5
512     35.5
513    314.5
514    207.5
Name: Additional_price, Length: 514, dtype: float64

**20.astype()** : ***To Convert the datatypes of column.***

In [102]:
df['Additional_price'].astype(float) # to change data type of column

1      1649.89
2       647.71
3       893.96
4      1505.41
5      1580.61
        ...   
510     323.13
511     828.61
512    1859.61
513     806.33
514    1167.56
Name: Additional_price, Length: 514, dtype: float64

**21.apply()** : ***Used to apply function to each row or column.***

In [108]:
df['Final_Report']=df['Country_Of_Origin'].apply(lambda x : "Eligible" if x =='India' else "not eligible")  # to apply function
df['Final_Report']

1      not eligible
2          Eligible
3      not eligible
4      not eligible
5          Eligible
           ...     
510        Eligible
511    not eligible
512        Eligible
513        Eligible
514    not eligible
Name: Final_Report, Length: 514, dtype: object

In [110]:
df['Final_Record']=df['Category'].apply(lambda x : "Good Product" if x == 'Dell' else "Better Product")  # to apply function
df['Final_Record']

1      Better Product
2      Better Product
3      Better Product
4      Better Product
5      Better Product
            ...      
510    Better Product
511    Better Product
512    Better Product
513    Better Product
514    Better Product
Name: Final_Record, Length: 514, dtype: object

**22.query()** : ***Use sql like quries to filter data.***

In [115]:
df.query("Company =='Apple' and Price > 1600") # apply query to filter data

Product_Name Company     Category    Price  Stock_Units Country_Of_Origin  \
44        Laptop   Apple  Electronics  1994.63        323.0           Germany   
58       Monitor   Apple  Electronics  1923.76        128.0           Germany   
60      Keyboard   Apple  Accessories  1949.19        220.0             China   
126      Monitor   Apple  Electronics  1906.44          5.0             India   
218       Webcam   Apple  Accessories  1797.81        454.0           Germany   
225        Mouse   Apple  Electronics  1681.31        291.0             India   
227      Printer   Apple  Accessories  1824.84        451.0                UK   
279       Router   Apple  Accessories  1795.80        337.0                UK   
303       Webcam   Apple  Peripherals  1804.85        361.0               USA   
359       Laptop   Apple  Electronics  1890.96        482.0             India   
365       Laptop   Apple  Accessories  1650.97        195.0                UK   
403       Laptop   Apple  Electronics  1873.26        273.0             India   
404        Mouse   Apple  Electronics  1839.67        394.0             China   
406     Keyboard   Apple  Peripherals  1937.78         27.0             Japan   
409       Laptop   Apple  Accessories  1693.36         78.0           Germany   
444        Mouse   Apple  Accessories  1636.42        166.0               USA   
448     Keyboard   Apple  Electronics  1717.59        194.0             Japan   
450   Headphones   Apple  Peripherals  1654.74        213.0             Japan   
495       Webcam   Apple  Peripherals  1679.25        364.0               USA   

    Launch_Date  Additional_price  Final_Report    Final_Record  
44   2023-12-30           1999.63  not eligible  Better Product  
58   2018-06-23           1928.76  not eligible  Better Product  
60   2019-02-03           1954.19  not eligible  Better Product  
126  2021-03-07           1911.44      Eligible  Better Product  
218  2021-06-09           1802.81  not eligible  Better Product  
225  2023-11-27           1686.31      Eligible  Better Product  
227  2022-08-08           1829.84  not eligible  Better Product  
279  2022-04-20           1800.80  not eligible  Better Product  
303  2023-04-12           1809.85  not eligible  Better Product  
359  2018-07-20           1895.96      Eligible  Better Product  
365  2018-01-15           1655.97  not eligible  Better Product  
403  2019-08-24           1878.26      Eligible  Better Product  
404  2019-06-03           1844.67  not eligible  Better Product  
406  2018-12-16           1942.78  not eligible  Better Product  
409  2020-03-28           1698.36  not eligible  Better Product  
444  2018-04-28           1641.42  not eligible  Better Product  
448  2022-03-31           1722.59  not eligible  Better Product  
450  2018-02-06           1659.74  not eligible  Better Product  
495  2021-08-05           1684.25  not eligible  Better Product

**23.pivot_table()** : ***Summarize data types of column aggregation function(sum,mean,median,mode).***

In [120]:
df.pivot_table(values="Price",index='Product_Name',columns='Company',aggfunc='sum') # to use aggregate function in pivot table

Company              Apple         Asus        Canon          Dell  \
Product_Name                                                         
Headphones     2697.900000  8292.738485  4529.050000   1047.960000   
Keyboard      12634.500000  2894.620000  4928.290000  10474.510000   
Laptop        20343.178485  6180.718485  9682.610000  12717.130000   
Monitor       12010.928485  4825.298485  1859.330000   5026.398485   
Mouse          9890.210000  3467.610000  8650.470000  10114.940000   
Printer       12807.680000  8907.290000  3387.210000   3752.490000   
Router         6845.030000  2258.650000  8115.600000   2679.490000   
Smartphone     2888.910000  9722.920000  5731.460000   5833.758485   
Tablet         2876.280000  8760.670000  2154.220000   6938.100000   
Webcam         7648.610000  5788.470000  9924.898485  11159.328485   

Company                HP       Lenovo      Logitech   Samsung  
Product_Name                                                    
Headphones    7619.100000  6632.540000   4900.818485  10960.54  
Keyboard      4170.220000  5757.368485   7342.928485   6414.93  
Laptop        8601.900000  2815.938485   4098.760000   3097.99  
Monitor       2667.688485  3718.850000   9447.148485  10545.48  
Mouse         7740.220000  8618.788485   4100.460000   3642.30  
Printer       6267.390000  7936.860000   4719.170000   6706.34  
Router        5095.080000  6091.380000    673.420000   1504.89  
Smartphone    8891.880000  9261.130000  11930.080000   3290.59  
Tablet        6619.450000  9626.958485   4503.880000   8693.99  
Webcam        9022.546970  2979.248485   4499.110000   6693.85

# Correlation

# **What is correlation ?**
- Correlation refers to the relationship between two or more variables.
- It measure how changes in one variable are linked to the changes in another variable.
- It apply  only numerical columns.

# **Type of Correlation**
- **Positive correlation :**
    - The value range is (0-1).it indicates two variables are positively correlated.i.e one variable is increase other variable is also increase or decrease.

- **Negative correlation :**
    - value range is(-1 to 0) . it indicates the two variable are negatively correlated. if one variable is increase the other variable is decrease.

- **Zero / No correlation :**
     - value range is 0. it indicates there is no relationship between variables.
    

In [130]:
correlation_table=df[['Price','Stock_Units','Additional_price']].corr()# to find correlation
print(correlation_table)

                    Price  Stock_Units  Additional_price
Price             1.00000     -0.00045           1.00000
Stock_Units      -0.00045      1.00000          -0.00045
Additional_price  1.00000     -0.00045           1.00000


- The price and Additional_price  are strongly correlated (positive correlated).
- The price and stock_units  are negatively correated.
- The stock_units and Additional_price has Zero/no correlation (there is no relation between this column).